crates/sui-types/src/messages_checkpoint.rs

```diff
@@ -119,10 +125,170 @@ impl Default for ECMHLiveObjectSetDigest {
+pub struct ObjectCheckpointState {
+    pub id: ObjectID,
+    pub digest: Option<ObjectDigest>,
+}
+
+impl ObjectCheckpointState {
+    pub fn new(id: ObjectID, digest: Option<ObjectDigest>) -> Self {
+        Self { id, digest }
+    }
+}
+
+impl From<ObjectChange> for ObjectCheckpointState {
+    fn from(object_change: ObjectChange) -> Self {
+        Self {
+            id: object_change.id,
+            digest: object_change.output_digest,
+        }
+    }
+}
+
+impl PartialEq for ObjectCheckpointState {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for ObjectCheckpointState {}
+
+impl PartialOrd for ObjectCheckpointState {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for ObjectCheckpointState {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.id.cmp(&other.id)
+    }
+}
+
+/// A sorted list of object checkpoint states (by object ID).
+#[derive(Debug)]
+pub struct ObjectCheckpointStates {
+    pub contents: Vec<ObjectCheckpointState>,
+}
+
+impl ObjectCheckpointStates {
+    pub fn new(contents: Vec<ObjectCheckpointState>) -> Self {
+        let mut contents = contents;
+        contents.sort();
+        Self { contents }
+    }
+
+    /// Compute the Merkle root of the object checkpoint states.
+    /// Assumes that the object checkpoint states are sorted by object ID.
+    pub fn digest(&self) -> SuiResult<ObjectCheckpointStatesDigest> {
+        let tree = MerkleTree::<Blake2b256>::build_from_unserialized(self.contents.iter())
+            .map_err(|e| SuiError::GenericAuthorityError {
+                error: format!("Failed to build Merkle tree: {}", e),
+            })?;
+        let root = tree.root().bytes();
+        Ok(ObjectCheckpointStatesDigest::from(root))
+    }
+}
+
+#[derive(Debug)]
+pub struct ObjectCheckpointStatesDigest {
+    pub digest: Digest,
+}
+
+impl From<[u8; 32]> for ObjectCheckpointStatesDigest {
+    fn from(digest: [u8; 32]) -> Self {
+        Self {
+            digest: Digest::new(digest),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CheckpointArtifacts {
+    pub latest_object_states: ObjectCheckpointStates,
+    // Add more.. e.g., execution digests, events, etc.
+}
+
+impl CheckpointArtifacts {
+    pub fn digest(&self) -> SuiResult<CheckpointArtifactsDigest> {
+        let latest_object_states_digest = self.latest_object_states.digest()?;
+
+        // When there are more artifacts, we can hash them all into a single digest.
+        Ok(CheckpointArtifactsDigest::from(
+            latest_object_states_digest.digest.into_inner(),
+        ))
+    }
+}
+
+impl From<&[&TransactionEffects]> for CheckpointArtifacts {
+    fn from(effects: &[&TransactionEffects]) -> Self {
+        let all_object_changes = effects
+            .iter()
+            .map(|e| e.object_changes())
+            .collect::<Vec<_>>();
```

Comment:
there seems to be no reason to create a vector here

Reply:
removed it